text/3502-cargo-script.md

```diff
@@ -0,0 +1,1229 @@
+    - With a `#!`, at minimum a local path must be passed in (e.g. `./` prefix) or the matching `PATH` element must be prefixed
+  - Might affect the quality of error messages for invalid subcommands unless we just assume
+  - Restricts access to more complex compiler settings unless a user switches
+    over to `cargo run` which might have different defaults (e.g. setting `RUST_BACKTRACE=1`)
+  - Forces us to have all commands treat these files equally (e.g.
+    `--<edition>` solution would need to be supported everywhere).
+  - Avoids the risk of overloading a `cargo-script`-like command to do
+    everything special for single-file packages, whether its running them,
+    expanding them into multi-file packages, etc.
+
+## First vs Third Party
+
+As mentioned, a reason for being first-party is to standardize the convention
+for this which also allows greater interop.
+
+A default implementation ensures people will use it.  For example, `clap`
+received an issue with a reproduction case using a `cargo-play` script that
+went unused because it just wasn't worth installing yet another, unknown tool,
+and it was unclear if it was interoperable with `rust-script`.
+
+This also improves the overall experience as you do not need the third-party
+command to replicate support for every potential feature including:
+- `cargo test` and other built-in cargo commands
+- `cargo expand` and other third-party cargo commands
+- `rust-analyzer` and other editor/IDE integration
+
+While other third-party cargo commands might not immediately adopt single-file
+packages, first-party support for them will help encourage their adoption.
+
+This still leaves room for third-party implementations, either differentiating themselves or experimenting with
+- Alternative caching mechanisms for lower overhead
+- Support for implicit `main`, like doc-comment examples
+- Template support for implicit `main` for customizing `use`, `extern`, `#[feature]`, etc
+- Short-hand dependency syntax (e.g. `//# serde_json = "*"`)
+- Prioritizing other workflows, like runtime performance
+
+## File association on Windows
+
+We could add a non-default association to run the file.
+We don't want it to be a default,
+to avoid unintended harm and due to the likelihood someone is going to want to
+edit these files.
+
+## File extension
+
+Should these files use `.rs` or a custom file extension?
+
+Reasons for a unique file type
+- Semantics are different than a normal `.rs` file
+  - Except already a normal `.rs` file has context-dependent semantics (`build.rs`, `main.rs`, `random.rs`),
+    so this doesn't seem too far off
+- Different file associations for Windows
+- Better detection by tools for the new semantics (particularly `rust-analyzer`)
+
+Downsides to a custom extension
+- Limited support by different tools (rust-analyzer, syntax highlighting, non-LSP editor actions) as adoptin rolls out
```

Comment:
```suggestion
- Limited support by different tools (rust-analyzer, syntax highlighting, non-LSP editor actions) as adoption rolls out
```

Reply:
IMHO the same is also true if we use `.rs`, since all of those tools currently don't know how to handle the frontmatter part of the script file

Reply:
Parsers and tools tend to be forgiving of syntax.

Reply:
Also, the frontmatter is optional while an extension wouldn't be